driver/src/test/java/org/neo4j/driver/v1/integration/TransactionBoltV3IT.java

```diff
@@ -0,0 +1,177 @@
+import org.neo4j.driver.v1.Session;
+import org.neo4j.driver.v1.StatementResult;
+import org.neo4j.driver.v1.StatementResultCursor;
+import org.neo4j.driver.v1.Transaction;
+import org.neo4j.driver.v1.TransactionConfig;
+import org.neo4j.driver.v1.Value;
+import org.neo4j.driver.v1.exceptions.TransientException;
+import org.neo4j.driver.v1.util.SessionExtension;
+
+import static java.time.Duration.ofSeconds;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsString;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.neo4j.driver.internal.util.Neo4jFeature.BOLT_V3;
+import static org.neo4j.driver.v1.util.TestUtil.await;
+
+@EnabledOnNeo4jWith( BOLT_V3 )
+class TransactionBoltV3IT
+{
+    @RegisterExtension
+    static final SessionExtension session = new SessionExtension();
+
+    @Test
+    void shouldSetTransactionMetadata()
+    {
+        Map<String,Object> metadata = new HashMap<>();
+        metadata.put( "key1", "value1" );
+        metadata.put( "key2", 42L );
+        metadata.put( "key3", false );
+
+        TransactionConfig config = TransactionConfig.builder()
+                .withMetadata( metadata )
+                .build();
+
+        try ( Transaction tx = session.beginTransaction( config ) )
+        {
+            tx.run( "RETURN 1" ).consume();
+
+            verifyTransactionMetadata( metadata );
+        }
+    }
+
+    @Test
+    void shouldSetTransactionMetadataAsync()
+    {
+        Map<String,Object> metadata = new HashMap<>();
+        metadata.put( "hello", "world" );
+        metadata.put( "key", ZonedDateTime.now() );
+
+        TransactionConfig config = TransactionConfig.builder()
+                .withMetadata( metadata )
+                .build();
+
+        CompletionStage<Transaction> txFuture = session.beginTransactionAsync( config )
+                .thenCompose( tx -> tx.runAsync( "RETURN 1" )
+                        .thenCompose( StatementResultCursor::consumeAsync )
+                        .thenApply( ignore -> tx ) );
+
+        Transaction transaction = await( txFuture );
+        try
+        {
+            verifyTransactionMetadata( metadata );
+        }
+        finally
+        {
+            await( transaction.rollbackAsync() );
+        }
+    }
+
+    @Test
+    void shouldSetTransactionTimeout()
+    {
+        // create a dummy node
+        session.run( "CREATE (:Node)" ).consume();
+
+        try ( Session otherSession = session.driver().session() )
+        {
+            try ( Transaction otherTx = otherSession.beginTransaction() )
+            {
+                // lock dummy node but keep the transaction open
+                otherTx.run( "MATCH (n:Node) SET n.prop = 1" ).consume();
+
+                TransactionConfig config = TransactionConfig.builder()
+                        .withTimeout( ofSeconds( 1 ) )
```

Comment:
does it need 1s? Can we give it 1ms to speed up this test? This 1s might slow down my fancy new compute to run tests. (The minimal legal value server accept is 1ms)

Similar for the async one.